src/main/java/co/thatch/example/metrics/InMemoryMetricsService.java

```diff
@@ -0,0 +1,117 @@
+package co.thatch.example.metrics;
+
+import co.thatch.example.metrics.model.Metric;
+import co.thatch.example.metrics.model.MetricBucket;
+import io.micronaut.runtime.server.event.ServerStartupEvent;
+import io.micronaut.runtime.event.annotation.EventListener;
+import io.micronaut.scheduling.annotation.Async;
+
+import javax.inject.Singleton;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.DoubleAdder;
+import java.util.stream.LongStream;
+
+
+@Singleton
+public class InMemoryMetricsService implements MetricsService {
+    private static final int MAX_WINDOW_SIZE = 1_800;
+    private static final int MAX_WINDOW_SIZE_WITH_BUFFER = MAX_WINDOW_SIZE + 1;
+
+    private final DoubleAdder requestCounter = new DoubleAdder();
+    private final double[] requestCountRecordings = new double[MAX_WINDOW_SIZE_WITH_BUFFER];
+    private final HashMap<Long, MetricHolder> metrics = new HashMap<>();
+    private final PriorityQueue<Long> timestamps = new PriorityQueue<>();
+
+    @Async
+    @EventListener
+    public void onStartup(ServerStartupEvent event) throws InterruptedException {
+        while (true) {
+            requestCountRecordings[Math.toIntExact(Instant.now().getEpochSecond() % requestCountRecordings.length)] = requestCounter.doubleValue();
+            Thread.sleep(50);
+        }
+    }
+
+    @Override
+    public void incrementRequestCount(String key, Instant requestInstant) {
+//        CompletableFuture.runAsync(() -> {
+            requestCounter.add(1.0);
+            Long timestamp = requestInstant.getEpochSecond();
+            synchronized (this) {
+                MetricHolder metricsForTimestamp = metrics.get(timestamp);
```

Comment:
Since we don't really care about a per-endpoint/per-path metric any more, could you use something like the code in the event handler above to somplify tracking how many requests we have had in each second?

Reply:
I've now separated the two approaches in seperate implementations to unpick form each other.